### ISSUE TITLE
Pickling error when using UMAP with pynndescent in a Spark-based environment

### ISSUE THREAD

**Issue by candalfigomoro** [lmcinnes/umap#477]
When using UMAP (without pynndescent) in a Spark-based environment, it works fine.

Instead, when using UMAP with pynndescent, I get the following error:

```
_pickle.PicklingError: Failed in nopython mode pipeline (step: nopython mode backend)
Can't pickle <class 'collections.FlatTree'>: attribute lookup FlatTree on collections failed
```

**Comment by lmcinnes** [comment]
Hmm, this is a new one. I'll have to look into this but I can't promise any swift resolution.

**Comment by ekerazha** [comment]
> Hmm, this is a new one. I'll have to look into this but I can't promise any swift resolution.

Related to https://github.com/lmcinnes/pynndescent/issues/93 ?

**Comment by koustavmukherjee** [comment]
I am also running into the same issue in a Spark based environment.

**Comment by lmcinnes** [comment]
A number of pickling issues were resolved in pynndescent v0.5 and umap v0.5 now depends on that -- are you seeing the issue with the newer versions? If so it seems like something specific to the spark environment, which is a little more confusing.

**Comment by Osherz5** [comment]
> Running into the same issue in a spark environment with umap-learn = 0.5.1 and pynndescent = 0.5.2.
> @koustavmukherjee Did you find any solution?

Also having this problem with these exact versions

**Comment by lmcinnes** [comment]
Is it specifically complaining about the ``collections.FlatTree``? If so I'll look and see if there is some reason that is still potentially causing issues.

**Comment by koustavmukherjee** [comment]
> Running into the same issue in a spark environment with umap-learn = 0.5.1 and pynndescent = 0.5.2.
> @koustavmukherjee Did you find any solution?

No @simon-slowik unfortunately I didn't find a solution yet for working with more recent versions of umap-learn from within a spark based environment. For now, I resorted to an older version of umap-learn (0.4.6) and using it without pynndescent.

**Comment by simon-slowik** [comment]
@lmcinnes yes the error message is specifically about `collections.FlatTree`.

It seems, pinning both umap-learn and pynndescent to 0.5.0 fixes the issue.

**Comment by lmcinnes** [comment]
Thanks @simon-slowik . It is particularly interesting that pinning to 0.5 works. I'll see if I can get this fixed.

**Comment by kmurphy4** [comment]
I was running into this today, and I think I found the cause.  It looks like [`pyspark` is overwriting `collections.namedtuple` with their own function](https://github.com/apache/spark/blob/v3.0.1/python/pyspark/serializers.py#L385), but that function isn't correctly [setting the `__module__`](https://github.com/python/cpython/blob/3.8/Lib/collections/__init__.py#L458) field.

Here's a minimal reproduction:

```python
>>> # don't load `pyspark`
>>> import collections
>>> collections.namedtuple("n", [])
<class '__main__.n'>
```

```python
>>> import pyspark
>>> import collections
>>> collections.namedtuple("n", [])
<class 'collections.n'>
```

I think there are a couple options for "fixing" this:
1. Fix the issue in `pyspark`: either set the `__module__` correctly (as the stdlib does) or "un-hijack" `collection.namedtuple` after they're done with it.
1. Pass the `module=` kwarg to `namedtuple` as `__name__`.  This avoids needing to make a change to `pyspark` but does require a change to `pynndescent` (seems like @lmcinnes is the maintainer of that one too, so it shouldn't be too hard :smile: ).  For example:
    ```python
    >>> import pyspark
    >>> import collections
    >>> collections.namedtuple("n", [], module=__name__)
    <class '__main__.n'>
    ```

In the meantime, users can monkey-patch `pynndescent.rp_trees.FlatTree` after `import`ing it to get around the bug:
```python
>>> import pyspark
>>> import pynndescent
>>> pynndescent.rp_trees.FlatTree.__module__  = "pynndescent.rp_trees"
>>> ... # my code goes after this
```
This is definitely the grossest solution, but it should work until one of the better solutions gets merged & released.

@lmcinnes if you think (2) is a reasonable course of action, I'm happy to open a PR.

**Comment by jamslaugh** [comment]
Sorry to bring this one back, but I am having exactly the same issue with pyspark on colab. I am performing an hyperoptimization on UMAP and this happens.

I use 

```python3
>>> import pyspark
>>> import pynndescent
>>> pynndescent.rp_trees.FlatTree.__module__  = "pynndescent.rp_trees"
```

then in `fmin` from `hyperopt` I set the following argument:

```python3
>>> trials=hyperopt.SparkTrials(n_trials)
```

and the error follows. Please note that if `n_trials` are `None` or a negative number, it seems to work properly for some trials, otherwise it spits back this error.

I'd like also to understand how can I choose to deactivate pynndescent from UMAP.

Thank you in advance.

**Comment by Wh1isper** [comment]
> Sorry to bring this one back, but I am having exactly the same issue with pyspark on colab. I am performing an hyperoptimization on UMAP and this happens.
> 
> I use
> 
> ```python
> >>> import pyspark
> >>> import pynndescent
> >>> pynndescent.rp_trees.FlatTree.__module__  = "pynndescent.rp_trees"
> ```
> 
> then in `fmin` from `hyperopt` I set the following argument:
> 
> ```python
> >>> trials=hyperopt.SparkTrials(n_trials)
> ```
> 
> and the error follows. Please note that if `n_trials` are `None` or a negative number, it seems to work properly for some trials, otherwise it spits back this error.
> 
> I'd like also to understand how can I choose to deactivate pynndescent from UMAP.
> 
> Thank you in advance.
> 
> BTW the error is:
> 
> Error Message
> edit 2:
> 
> if I run umap as it is, it gives me no issues. I think this must be a problem between hyperopt, spark and umap.

Same, but I'm not sure hyperopt used in my project.

I speculate if it has something to do with pyspark's optimization of CPickle serialization in python < 3.8 environment?https://github.com/apache/spark/blob/f84018a4810867afa84658fec76494aaae6d57fc/python/pyspark/serializers.py#L361

I am using pyspark 3.1.2 and umap-learn 0.5.2, python 3.7.10
